docker-base-build/install_pinned.py

```diff
@@ -0,0 +1,377 @@
+
+
+def version_from_requirement(requirement: Requirement) -> str:
+    if requirement.url is not None:
+        raise ValueError('Cannot get version from an URL requirement')
+    pin = None
+    for spec in requirement.specifier:
+        if spec.operator in {'==', '==='} and '*' not in spec.version:
+            pin = spec.version
+    if pin is None:
+        raise ValueError(f'No version pinned for {requirement.name}')
+    for spec in requirement.specifier:
+        if pin not in spec:
+            raise ValueError(f'{requirement.name}: pinned version {pin} does not satisfy {spec}')
+    return pin
+
+
+def evaluate_marker(requirement: Requirement, parent: Requirement) -> bool:
+    """Evaluate whether `requirement` should be applied.
+
+    This handles the special behaviour of ``extra`` in the marker spec, which
+    is used in wheels to reference the extras being installed for the
+    depending package (see PEP 508).
+    """
+    if requirement.marker is None:
+        return True
+    try:
+        return requirement.marker.evaluate()
+    except UndefinedEnvironmentName:
+        extras = parent.extras or ['']
+        for extra in extras:
+            if requirement.marker.evaluate({'extra': extra}):
+                return True
+        return False
+
+
+def get_dependencies(requirement: Requirement) -> Sequence[Requirement]:
+    with tempfile.TemporaryDirectory() as cache_dir:
+        # Pip uses a vendored version of packaging, so we have to translate
+        req = PipRequirement(str(requirement))
+        ireq = InstallRequirement(req, None)
+        pypi = piptools.repositories.PyPIRepository([], cache_dir)
+        # Map from InstallRequirement back to packaging Requirement
+        deps = [Requirement(str(r.req)) for r in pypi.get_dependencies(ireq)]
+        deps = [dep for dep in deps if evaluate_marker(dep, requirement)]
+        for dep in deps:
+            dep.name = canonicalize_name(dep.name)
+            # We've checked it, and 'extra' markers can cause problems later
+            # as we don't have the context
+            dep.marker = None
+        return deps
+
+
+def resolve(items: Iterable[Union[Package, str]]) -> Sequence[Union[Requirement, str]]:
+    def add_constraint(pkg: Package):
+        name = pkg.requirement.name
+        if name in constraints:
+            constraints[name] = merge_packages(constraints[name], pkg)
+        else:
+            constraints[name] = pkg
+        return constraints[name]
+
+    options = []
+    constraints: Dict[str, Package] = {}
+    install: Dict[str, Requirement] = {}
+    q: Deque[Requirement] = deque()
+    for item in items:
+        if isinstance(item, str):
+            options.append(item)
+        elif item.constraint:
+            add_constraint(item)
+        else:
+            q.append(item.requirement)
+
+    errors = []
+    while q:
+        req = q[0]
+        q.popleft()
```

Comment:
Why not `req = q.popleft()`?

Reply:
Because I've spent too much of my life using C++. Done.